README.md
Raspi IO
========

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/nebrius/raspi-io?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Raspi IO is an I/O plugin for the [Johnny-Five](https://github.com/rwaldron/johnny-five) Node.js robotics platform that enables Johnny-Five to control the hardware on a Raspberry Pi. The API docs for this module can be found on the [Johnny-Five I/O Plugin specification page](https://github.com/rwaldron/io-plugins), except for the constructor which is documented below. Raspi IO supports all models of the Raspberry Pi, except for the Model A and compute modules.

If you have a bug report, feature request, or wish to contribute code, please be sure to check out the [Contributing Guide](/CONTRIBUTING.md).

## System Requirements

- Raspberry Pi Model B Rev 1 or newer (sorry Model A users), except compute modules
- Raspbian Jessie or newer
  - [Node-RED](http://nodered.org/) works, but can be finicky and difficult to debug.
  - See https://github.com/nebrius/raspi-io/issues/24 for more info about support for other OSes
- Node 6.4.0 or newer

Detailed instructions for getting a Raspberry Pi ready for NodeBots, including how to install Node.js, can be found in the [wiki](https://github.com/nebrius/raspi-io/wiki/Getting-a-Raspberry-Pi-ready-for-NodeBots)

**Warning:** this module must be installed as a normal user, but run as the root user

## Installation

Install with npm on a Raspberry Pi (not a laptop):

```Shell
npm install raspi-io
```

**Note:** You must reboot your Raspberry Pi the first time you install Raspi IO on it!

## Usage

Using Raspi IO inside of Johnny-Five is fairly straightforward, although does take an extra step compared to using Johnny-Five on the Arduino Uno:

```JavaScript
const Raspi = require('raspi-io').RaspiIO;
const five = require('johnny-five');
const board = new five.Board({
  io: new Raspi()
});

board.on('ready', () => {

  // Create an Led on pin 7 (GPIO4) on P1 and strobe it on/off
  // Optionally set the speed; defaults to 100ms
  (new five.Led('P1-7')).strobe();

});
```

Then, run your code with:

```bash
sudo node index.js
```

The `io` property must be specified explicitly to differentiate from trying to control, say, an Arduino Uno that is plugged into the Raspberry Pi. Note that we specify the pin as `"P1-7"`, not `7`. See the [section on pins](#pin-naming) below for an explanation of the pin numbering scheme on the Raspberry Pi, which can be a bit confusing.

**Warning:** this module _must_ be run as root, even though it cannot be installed as root.

**Note:** This module is not intended to be used directly. If you do not want to use Johnny-Five, I recommend taking a look at [Raspi.js](https://github.com/nebrius/raspi), which underpins this library and is a little more straight-forward to use than using Raspi IO directly.

## Important Notes

### Pin Naming

The pins on the Raspberry Pi are a little complicated. There are multiple headers on some Raspberry Pis with extra pins, and the pin numbers are not consistent between Raspberry Pi board versions.

To help make it easier, you can specify pins in three ways. The first is to specify the pin by function, e.g. `'GPIO18'`. The second way is to specify by pin number, which is specified in the form "P[header]-[pin]", e.g. `'P1-7'`. The final way is specify the [Wiring Pi virtual pin number](http://wiringpi.com/pins/), e.g. `7`. If you specify a number instead of a string, it is assumed to be a Wiring Pi number.

Be sure to read the [full list of pins](https://github.com/nebrius/raspi-io/wiki/Pin-Information) on the supported models of the Raspberry Pi.

### I2C notes

There are a few limitations and extra steps to be aware of when using I2C on the Raspberry Pi.

First, note that the I2C pins can _only_ be used for I2C with Raspi IO, even though they are capable of GPIO at the hardware level.

Also note that you will need to edit `/boot/config.txt` in order to change the I2C baud rate from the default, if you need to. If you notice that behavior is unstable while trying to communicate with another microcontroller, try setting the baudrate to 10000 from the default 100000. This instability has been observed on the Arduino Nano before.

After you install Raspi IO for the first time, you _must_ reboot your Raspberry Pi. I2C support is not enabled by default, and this module runs a script to enable it automatically and adjust a few I2C settings. These settings will not take effect until you reboot your Pi.

Finally, if you try to access a device that doesn't exist, you will get an error stating `EIO, i/o error` (sorry it's not very descriptive, but it comes from the operating system and I can't change it).

### Serial notes

There are also a few limtations and extra steps to be aware of when using Serial (UART) on the Raspberry Pi.

As with I2C, the serial pins can _only_ be used for serial with Raspi IO, even though they are capable of GPIO at the hardware level.

**If you _are not_ running a Raspberry Pi without WiFi:**

All older versions of the Raspberry Pi enable a TTY console over serial, meaning that you can use the `screen` command on *NIX computers to log in to the Raspberry Pi over serial. This can get in the way of using the serial port for robotics, however. To disable TTY over serial, do the following:

1. Run `sudo raspi-config` to start the Raspberry Pi configuration utility
2. Select `5 Interface Options`
3. Select `P6 Serial Options`
4. Select `No` when asked `Would you like a login shell to be accessible over serial?`
5. Select `Yes` when asked `Would you like the serial port hardware to be enabled?`
6. Select `OK`
7. Select `Finish` and select `Yes` to reboot when prompted

**WARNING: If you _are_ running a Raspberry Pi with WiFi:**

The Bluetooth module on these Raspberry Pis is controlled using the serial port, meaning it cannot be used directly while also using Bluetooth. Using this module with the default serial port will _disable_ bluetooth!

For an in-depth discussion on why and how to work around it, read https://raspberrypi.stackexchange.com/questions/45570/how-do-i-make-serial-work-on-the-raspberry-pi3.

## API

### new RaspiIO(options)

Instantiates a new Raspi IO instance with the given options

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>options (optional)</td>
    <td>Object</td>
    <td>The configuration options.</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Property</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>enableSerial (optional)</td>
          <td>boolean</td>
          <td>Enables the use of the serial port by Johnny-Five. The default is <code>true</code>, except on models with Bluetooth where it is <code>false</code> (currently the Raspberry Pi 3 Model B, Raspberry Pi 3 Model B+, and Raspbery Pi Zero W)</td>
        </tr>
        <tr>
          <td>enableI2C (optional)</td>
          <td>boolean</td>
          <td>Enables the use of the I2C port by Johnny-Five. The default is <code>true</code></td>
        </tr>
        <tr>
          <td>includePins (optional)</td>
          <td>Array&lt;Number|String&gt;</td>
          <td>A list of pins to include in initialization. If specified, any pins <em>not</em> listed here will not be initialized or available for use by Raspi IO. If not specified, all pins will be available for use by Raspi IO.</td>
        </tr>
        <tr>
          <td>excludePins (optional)</td>
          <td>Array&lt;Number|String&gt;</td>
          <td>A list of pins to exclude from initialization. If specified, any pins listed here will not be initialized or available for use by Raspi IO. If not specified, all pins will be available for use by Raspi IO.</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

License
=======

The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
## 11.0.0 (2020-1-10)

- Updated dependencies
- BREAKING CHANGE: dropped support for end-of-lifed versions of Node.js, so this now only supports Node 10, 12, and 13

## 10.0.4 (2019-10-5)

- Added `arm64` to the `cpu` field in package.json for installing on 64-bit Raspberry Pis

## 10.0.3 (2019-9-12)

- Fixed a bug where serial wasn't being disabled on the RPi 4

## 10.0.2 (2019-4-15)

- Fixed a bug where GPIO pins were not marked as PWM pins correctly

## 10.0.1 (2019-2-21)

- Brought in an updated dependency

## 10.0.0 (2019-2-21)

- BREAKING CHANGE: Under the hood we now use the PWM clock instead of the PCM clock for software PWM.
    - You can no longer use hardware PWM with Raspi IO (which Raspi IO itself does not support)
    - You now _can_ use I2S in conjunction with Raspi IO
- BREAKING CHANGE: there is no longer a default export. Instead use `const Raspi = require('raspi-io').RaspiIO` or `import { RaspiIO as Raspi } from 'raspi-io';` to get the old behavior to work
- Added a new constructor property `enableI2C` that allows users to programatically enable or disable I2C
- BREAKING CHANGE: The `name` property of this module now returns `"Raspi IO"` instead of `"RasperryPi-IO"` to conform with naming conventions elsewhere
- BREAKING CHANGE: removed the `enableSoftPwm` and switched to using software PWM by default for all pins
    - Note: this means that hardware PWM is not used on the pins that support hardware PWM, we still use software PWM
- BREAKING CHANGE: `digital-read-${pin}` event names are now normalized. If, for example, you called `digitalRead("GPIO18", () => {})` on the Raspberry Pi, before the event name would be `digital-read-GPIO18`, but now it's `digital-read-1`

Under the hood, I switched from `raspi-io-core`@2.x.x to `j5-io`@3.x.x. Note: J5 IO is a major rewrite of Raspi IO Core. Check out [J5 IO's changelog](https://github.com/nebrius/core-io/blob/master/CHANGELOG.md) for details. Note that many of those breaking changes are not surfaced in Raspi IO, and all breaking changes are listed above.


## 9.0.1 (2018-12-10)

- Updated raspi-board to the latest version, which includes breaking changes. These changes are not surfaced by Raspi IO.

## 9.0.0 (2018-4-21)

- Updated dependencies
- BREAKING CHANGE: Dropped support for Node.js 4

## 8.1.1 (2018-4-02)

- Pulled in latest version of raspi-io-core to switch servo methods to using range instead of degrees

## 8.1.0 (2017-9-12)

- Enforced support for Node.js < 4.0.0 via package.json "engines" field.
    - This does not change the actual supported platforms for Raspi IO, just makes it more explicit
- Dropped support for attempting to install on non-arm platforms via package.json's "cpu" field.
    - Attempting to install this on a non-Raspberry Pi platform before gave a bunch of obtuse errors, so this doesn't _actually_ change the ability to install raspi-io, but does make it fail earlier, harder, and more obviously
- Updated dependencies that had breaking changes in prep for adding 1-wire support.
    - The breaking changes in dependencies are not surfaced as breaking changes in Raspi IO itself.

## 8.0.1(2017-4-29)

- Pulled in a dependency change that fixes a bug in servo duty cycle calculation

## 8.0.0 (2017-4-23)

- Updated dependencies that gets rid of Wiring Pi in favor of pigpio.
  - POTENTIALLY BREAKING CHANGE: there are no _known_ breakages with this update, but there are a lot of under the hood changes, and better safe than sorry.
- Disabled serial on the Raspberry Pi Zero W by default because it suffers from the same serial limitations as the Raspberry Pi 3

## 7.4.0 (2017-4-19)

- Added proper Raspberry Pi Zero W support, which sets `enableSerial` to false

## 7.3.1 (2017-4-7)

- Dependency bump

## 7.3.0 (2017-4-7)

- Added `enableSerial` property to get around crashing on the Raspberry Pi 3

## 7.2.4 (2017-2-17)

- Updated raspi-io-core to bring in better error messaging around pin modes

## 7.2.3 (2017-22-1)

- Publishing a new version to update the README on npmjs.com
- Removed outdated and broken unit tests, they weren't even remotely comprehensive anyways.

## 7.2.2 (2017-1-9)

- Updated dependencies to pull in a bug fix

## 7.2.1 (2017-1-5)

- Bumped the raspi-gpio dependency. There is a potentially breaking change in raspi-gpio, but because of how this module uses it, there is no breaking change here.

## 7.2.0 (2016-12-30)

- Split off almost all functionality of this module into [Raspi IO Core](https://github.com/nebrius/raspi-io-core)
  - Note: there should be no functionality change from this package, but giving it a minor version bump anyways
  - By splitting this off, it should be possible to create variants of raspi-io that target different OSes and configurations, e.g. Windows 10 IoT Core

## 7.1.0 (2016-11-29)

- Added support for software-based PWM

## 7.0.0 (2016-11-25)

- Fixed a bug where writing 0 to GPIO pins on startup wasn't actually writing 0
- Updated dependencies
  - POTENTIALLY BREAKING CHANGE: removed dependency on raspi-wiringpi because Raspbian now ships with Wiring Pi by default. This _may_ break other OSes
  - POTENTIALLY BREAKING CHANGE: I2C introduced a breaking change around write size limits, see the [raspi-i2c changelog](https://github.com/nebrius/raspi-i2c/blob/master/CHANGELOG.md) for more info.

## 6.2.0 (2016-11-22)

- Added the ability to specify pins to include or exclude in initialization

## 6.1.1 (2016-10-29)

- Updated dependencies to bring in new functionality
- Fixed a bug where `i2cConfig` treated the delay as milliseconds instead of seconds
- Added support for enabling pull down resistors

## 6.1.0 (2016-7-7)

- Switched dependency ranges to ^
- Bumped dependencies to bring in support for a new Raspberry Pi Zero revision

## 6.0.1 (2016-3-20)

- Fixed a bug where unknown pins threw the error 'Unknown pin "null"', which wasn't very helpful

## 6.0.0 (2016-3-20)

- Added support for the `serial*` methods

lib/board.js
/*
Copyright (c) 2014 Bryan Hughes <bryan@theoreticalideations.com>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the 'Software'), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
*/

var fs = require('fs');
var revision;
var boardPins;
var boardPinAliases;

// Version lookup info obtained from:
// http://raspberryalphaomega.org.uk/2013/02/06/automatic-raspberry-pi-board-revision-detection-model-a-b1-and-b2/
var BOARD_REVISIONS = {
  0x00: 'Unknown',
  0x02: 'B1',
  0x03: 'B1',
  0x04: 'B2',
  0x05: 'B2',
  0x06: 'B2',
  0x07: 'A',
  0x08: 'A',
  0x09: 'A',
  0x0D: 'B2',
  0x0E: 'B2',
  0x0F: 'B2',
  0x10: 'B+'
};

var modes = module.exports.modes = Object.freeze({
  INPUT: 0,
  OUTPUT: 1,
  ANALOG: 2,
  PWM: 3,
  SERVO: 4
});

// Pin function info obtained from:
// http://elinux.org/Rpi_Low-level_peripherals
function createPinDescriptor(gpio) {

  // Calculate the supported modes
  var supportedModes = [];
  if (typeof gpio === 'number') {
    supportedModes.push(modes.INPUT, modes.OUTPUT);
  }

  // Return the descriptor
  return {
    supportedModes: supportedModes,
    mode: modes.INPUT,
    get value() {
      return this._instance.readValueSync();
    },
    set value(value) {
      if (this.mode != modes.OUTPUT) {
        throw new Error('Cannot write to pin because it is not in output mode');
      }
      return this._instance.writeValueSync(value);
    },
    analogChannel: 127,
    _gpio: gpio
  };
}
var B1_BOARD_PINS = [

  // There's not really a pin 0, so we create an empty entry
  createPinDescriptor(),

  // 1-10
  createPinDescriptor(),
  createPinDescriptor(),
  createPinDescriptor(0),
  createPinDescriptor(),
  createPinDescriptor(1),
  createPinDescriptor(),
  createPinDescriptor(4),
  createPinDescriptor(14),
  createPinDescriptor(),
  createPinDescriptor(15),

  // 11-20
  createPinDescriptor(17),
  createPinDescriptor(18),
  createPinDescriptor(21),
  createPinDescriptor(),
  createPinDescriptor(22),
  createPinDescriptor(23),
  createPinDescriptor(),
  createPinDescriptor(24),
  createPinDescriptor(10),
  createPinDescriptor(),

  // 21-26
  createPinDescriptor(9),
  createPinDescriptor(25),
  createPinDescriptor(11),
  createPinDescriptor(8),
  createPinDescriptor(),
  createPinDescriptor(7)
];
var B1_BOARD_PIN_ALIASES = {
  GPIO0: 3,
  GPIO1: 5,
  GPIO4: 7,
  GPIO14: 8,
  GPIO15: 10,
  GPIO17: 11,
  GPIO18: 12,
  GPIO21: 13,
  GPIO22: 15,
  GPIO23: 16,
  GPIO24: 18,
  GPIO10: 19,
  GPIO9: 21,
  GPIO25: 22,
  GPIO11: 23,
  GPIO8: 24,
  GPIO7: 26,

  SDA: 3,
  SCL: 5,
  GPCLK0: 7,
  TXD: 8,
  RXD: 10,
  PCM_CLK: 12,
  MOSI: 19,
  MISO: 21,
  SCLK: 11,
  CE0: 24,
  CE1: 26
};
var B2_BOARD_PINS = [

  // There's not really a pin 0, so we create an empty entry
  createPinDescriptor(),

  // 1-10
  createPinDescriptor(),
  createPinDescriptor(),
  createPinDescriptor(2),
  createPinDescriptor(),
  createPinDescriptor(3),
  createPinDescriptor(),
  createPinDescriptor(4),
  createPinDescriptor(14),
  createPinDescriptor(),
  createPinDescriptor(15),

  // 11-20
  createPinDescriptor(17),
  createPinDescriptor(18),
  createPinDescriptor(27),
  createPinDescriptor(),
  createPinDescriptor(22),
  createPinDescriptor(23),
  createPinDescriptor(),
  createPinDescriptor(24),
  createPinDescriptor(10),
  createPinDescriptor(),

  // 21-26
  createPinDescriptor(9),
  createPinDescriptor(25),
  createPinDescriptor(11),
  createPinDescriptor(8),
  createPinDescriptor(),
  createPinDescriptor(7)
];
var B2_BOARD_PIN_ALIASES = {
  GPIO2: 3,
  GPIO3: 5,
  GPIO4: 7,
  GPIO14: 8,
  GPIO15: 10,
  GPIO17: 11,
  GPIO18: 12,
  GPIO27: 13,
  GPIO22: 15,
  GPIO23: 16,
  GPIO24: 18,
  GPIO10: 19,
  GPIO9: 21,
  GPIO25: 22,
  GPIO11: 23,
  GPIO8: 24,
  GPIO7: 26,

  SDA: 3,
  SCL: 5,
  GPCLK0: 7,
  TXD: 8,
  RXD: 10,
  PCM_CLK: 12,
  MOSI: 19,
  MISO: 21,
  SCLK: 11,
  CE0: 24,
  CE1: 26
};
var BPLUS_BOARD_PINS = [

  // There's not really a pin 0, so we create an empty entry
  createPinDescriptor(),

  // 1-10
  createPinDescriptor(),
  createPinDescriptor(),
  createPinDescriptor(2),
  createPinDescriptor(),
  createPinDescriptor(3),
  createPinDescriptor(),
  createPinDescriptor(4),
  createPinDescriptor(14),
  createPinDescriptor(),
  createPinDescriptor(15),

  // 11-20
  createPinDescriptor(17),
  createPinDescriptor(18),
  createPinDescriptor(27),
  createPinDescriptor(),
  createPinDescriptor(22),
  createPinDescriptor(23),
  createPinDescriptor(),
  createPinDescriptor(24),
  createPinDescriptor(10),
  createPinDescriptor(),

  // 21-30
  createPinDescriptor(9),
  createPinDescriptor(25),
  createPinDescriptor(11),
  createPinDescriptor(8),
  createPinDescriptor(),
  createPinDescriptor(7),
  createPinDescriptor(),
  createPinDescriptor(),
  createPinDescriptor(5),
  createPinDescriptor(),

  // 31-40
  createPinDescriptor(6),
  createPinDescriptor(12),
  createPinDescriptor(13),
  createPinDescriptor(),
  createPinDescriptor(19),
  createPinDescriptor(16),
  createPinDescriptor(26),
  createPinDescriptor(20),
  createPinDescriptor(),
  createPinDescriptor(21)
];
var BPLUS_BOARD_PIN_ALIASES = {
  GPIO2: 3,
  GPIO3: 5,
  GPIO4: 7,
  GPIO14: 8,
  GPIO15: 10,
  GPIO17: 11,
  GPIO18: 12,
  GPIO27: 13,
  GPIO22: 15,
  GPIO23: 16,
  GPIO24: 18,
  GPIO10: 19,
  GPIO9: 21,
  GPIO25: 22,
  GPIO11: 23,
  GPIO8: 24,
  GPIO7: 26,
  GPIO5: 29,
  GPIO6: 31,
  GPIO12: 32,
  GPIO13: 33,
  GPIO19: 35,
  GPIO16: 36,
  GPIO26: 37,
  GPIO20: 38,
  GPIO21: 40,

  SDA: 3,
  SCL: 5,
  GPCLK0: 7,
  TXD: 8,
  RXD: 10,
  PCM_CLK: 12,
  MOSI: 19,
  MISO: 21,
  SCLK: 11,
  CE0: 24,
  CE1: 26
};

// Set up support for the status LED
fs.writeFileSync('/sys/class/leds/led0/trigger', 'none');
var led0 = B1_BOARD_PINS.led0 = B2_BOARD_PINS.led0 = {
  supportedModes: [ modes.INPUT, modes.OUTPUT ],
  mode: modes.OUTPUT,
  get value() {
    return parseInt(fs.readFileSync('/sys/class/leds/led0/brightness').toString()) ? 1 : 0;
  },
  set value(value) {
    if (this.mode != modes.OUTPUT) {
      throw new Error('Cannot write to pin because it is not in output mode');
    }
    if (value !== this.value) {
      fs.writeFileSync('/sys/class/leds/led0/brightness', value ? '255' : '0');
    }
  },
  analogChannel: 127,
  _gpio: 'led0',
  _instance: {
    init: function(cb) {
      cb();
    },
    watchValue: function(handler) {}, // Ignored
    setDirectionSync: function() {}, // Ignored
    writeValueSync: function(value) {
      led0.value = value;
    },
    readValueSync: function() {
      return led0.value;
    }
  }
};

module.exports.getBoardVersion = function(cb) {
  if (revision) {
    process.nextTick(cb.bind(undefined, null, revision));
    return;
  }
  fs.readFile('/proc/cpuinfo', function(err, buf) {
    if (err) {
      cb(err);
    } else {
      var rev = buf.toString().match(/Revision\s*:\s*(.*)/);
      rev = parseInt(rev && rev[1] || '0', 16);
      revision = BOARD_REVISIONS[rev];

      switch(revision) {
        case 'A':
          // Information is scarce, going to have to reverse-engineer the schematics to figure out pinouts
          throw new Error('Rev A boards are not yet supported.');
        case 'B1':
          boardPins = B1_BOARD_PINS;
          boardPinAliases = B1_BOARD_PIN_ALIASES;
          break;
        case 'B2':
          boardPins = B2_BOARD_PINS;
          boardPinAliases = B2_BOARD_PIN_ALIASES;
          break;
        case 'B+':
          boardPins = BPLUS_BOARD_PINS;
          boardPinAliases = BPLUS_BOARD_PIN_ALIASES;
          break;
        default:
          throw new Error('Unknown board revision ' + revision);
      }

      cb(null, revision);
    }
  });
};

module.exports.getBoardPins = function(cb) {
  if (!boardPins) {
    module.exports.getBoardVersion(function (err) {
      if (err) {
        cb(err);
        return;
      }
      cb(null, boardPins, boardPinAliases);
    });
  } else {
    process.nextTick(cb.bind(undefined, null, boardPins, boardPinAliases));
  }
};
